### PR TITLE
*: switch to flat plan in stmt summary, slow log, etc

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/35888

### What is changed and how it works?

- Cache the plan and corresponding flat plan in the stmtctx. (adapter.go, stmtctx.go, ...)
- `GetStatsInfo` => `GetStatsInfoFromFlatPlan` (util.go)
- `GenHintsFromPhysicalPlan` => `GenHintsFromFlatPlan` (hints.go)
- `EncodePlan` => `EncodeFlatPlan` (encode.go)
- `NormalizePlan` => `NormalizeFlatPlan` (encode.go)
- Some minor changes.
  - Add some missing database names in the operator. (physical_plans.go, find_best_task.go, planbuilder.go)
  - Some function parameter types are changed. (codec.go)

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/encode.go [hunk 1]
```diff
@@ -40,6 +141,7 @@ type planEncoder struct {
 }
 
 // EncodePlan is used to encodePlan the plan to the plan tree with compressing.
+// Deprecated: FlattenPhysicalPlan() + EncodeFlatPlan() is preferred.
 func EncodePlan(p Plan) string {
```
Comment (winoros) [inline planner/core/encode.go:145]: Now it's only used in two benchmark tests and one true unit test, we can remove it actually?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -1071,6 +1071,7 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
 			Columns:         ds.Columns,
 			Table:           is.Table,
 			TableAsName:     ds.TableAsName,
+			DBName:          ds.DBName,
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go:1074]: It's missed and may output the wrong table name in the explain result?

File: planner/core/hints.go [hunk 3]
```diff
@@ -22,7 +22,44 @@ import (
...
+			p := op.Origin.(PhysicalPlan)
+			hints = genHintsFromSingle(p, nodeTp, hints)
+		}
+	}
+	return hints
+}
+
 // GenHintsFromPhysicalPlan generates hints from physical plan.
+// Deprecated: FlattenPhysicalPlan() + GenHintsFromFlatPlan() is preferred.
 func GenHintsFromPhysicalPlan(p Plan) []*ast.TableOptimizerHint {
```
Comment (winoros) [inline planner/core/hints.go:63]: The only place it is called in prod env is in the `Optimize`'s SPM evolve phase. In that place, we get the hints from the plan returned by `optmize()`. Would we also flatten the plan there?

File: planner/core/encode.go (hunk 1 above)
Comment (time-and-fate) [reply planner/core/encode.go:145]: Yes.
I'll make sure all those methods can be removed safely and remove them in future PRs.

File: planner/core/exhaust_physical_plans.go (hunk 2 above)
Comment (time-and-fate) [reply planner/core/exhaust_physical_plans.go:1074]: It's not quite related to this PR. It won't cause the wrong table name, just sometimes the `DBName` field is missing.
We don't output DB names in EXPLAIN, but I'll use them in the "access object" in the next PR. So I just fixed it here for convenience.

**Review by winoros — APPROVED** [review]

File: executor/adapter.go [hunk 4]
```diff
@@ -926,6 +928,11 @@ func (a *ExecStmt) FinishExecuteStmt(txnTS uint64, err error, hasMoreResults boo
 		a.Ctx.GetTxnWriteThroughputSLI().AddReadKeys(execDetail.ScanDetail.ProcessedKeys)
 	}
 	succ := err == nil
+	if a.Plan != nil {
+		// If this statement has a Plan, the StmtCtx.plan should have been set when it comes here,
+		// but we set it again in case we missed some code paths.
```
Comment (AilinKid) [inline executor/adapter.go:933]: 😂

**Review by AilinKid — APPROVED** [review]

**Pull request merged**